docs/hugo/content/design/ADR-2024-07-Dynamic-Export-To-Secret-Or-ConfigMap.md

```diff
@@ -0,0 +1,428 @@
+```yaml
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+```
+
+or if everything required a dynamic expression.
+
+**Pros**
+
+* Could combine multiple resources data into a single ConfigMap, though this brings questions of ownership
+  and deletion/propagation.
+
+**Cons**
+
+* Another way to do things. Possibly confusing because there will be 2 ways to export data to configmap/secret 
+  for many existing resources.
+* More complicated, especially setting up watches on N resource types so that changes are correctly monitored and
+  propagated quickly. That additional complexity doesn't necessarily bring all that many wins.
+* Access to secret data is more difficult to implement. We would need to have a static mapping of resource types and
+  the `Exporter` would need to call each referenced resources `ExportKubernetesResources`, and we'd likely need to 
+  refactor `ExportKubernetesResources` some to support the new pattern (though we may need to do this anyway).
+
+### Access to secret data
+
+Getting access to data to export to ConfigMaps is pretty easy, we'll just pass the whole `resource` to
+the expression. As shown elsewhere fields will be accessible in expressions like `resource.status.id`.
+
+Secret data is by definition _not_ on the `resource` though, and is often returned in a list format only after
+calling a special API. As the [requirements](#requirements) section discusses, we need to ensure that we only
+invoke this secret API if and when a user asks for secrets. Then, actual access to the secret fields should
+be off of a _separate_ object: `secret.hostName` or `secret.key1` for example.
+
+We currently have an extension for Secrets or custom ConfigMaps whose signature is:
+```go
+ExportKubernetesResources(
+    ctx context.Context,
+    obj genruntime.MetaObject,
+    armClient *genericarmclient.GenericClient,
+    log logr.Logger,
+) ([]client.Object, error)
+```
+
+This extension is used both by handcrafted extensions used to export secrets and autogenerated extensions to support
+`$generatedConfigs`.
+This implementation is not suitable to supporting the new approach because it returns `[]client.Object`, 
+but to supply the CEL-expressions we need a return type of an object or a map whose
+properties/keys line up with the secret names. Alternatively we could delegate evaluation of the CEL expression
+to the extension itself but this would require relatively complex code autogenerated for each type, since every
+type will support the new `operatorSpec.formattedConfigMaps` field.
+
+Since `ExportKubernetesResources` is already used for a variety of purposes, rather than change its signature I propose
+we add a new sister interface to it called `SecretGetter` for resources that have Secret APIs to call.
+This interface must be implemented manually alongside the manual implementation of the
+`KubernetesExporter` resource.
```

Comment:
If it must be implemented alongside - that is, if it's mandatory - why not extend the existing `KubernetesExporter` interface with a second method, to force this.

Reply:
Not every KubernetesExporter produces secrets. If all it produces is configmaps we don't need to have the secret specific one

Reply:
I've clarified this wording some and pointed out we could add the method to the existing interface, but I think a separate interface will ultimately be cleaner. Not married to making a decision here though we can also discuss during implementation.